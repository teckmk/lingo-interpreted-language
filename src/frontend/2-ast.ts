export type NodeType =
  // Statements
  | "Program"
  | "VarDeclaration"
  | "MultiVarDeclaration"
  | "FunctionDeclaration"
  | "ReturnStatement"
  | "IfElseStatement"
  | "WhileStatement"
  // Expressions
  | "AssignmentExpr"
  | "MemberExpr"
  | "CallExpr"
  | "BinaryExpr"
  | "FunctionParam"
  // Literals
  | "Property"
  | "ObjectLiteral"
  | "StringLiteral"
  | "NumericLiteral"
  | "Identifier"
  | "ArrayLiteral"
  | "DocComment"

export type Type = "string" | "number" | "bool" | "array" | "object" | "dynamic"
export type VarModifier = "constant" | "final" | "variable"

export interface Stmt {
  kind: NodeType
}

export interface Program extends Stmt {
  kind: "Program"
  body: Stmt[]
}

export interface AssignmentExpr extends Expr {
  kind: "AssignmentExpr"
  assigne: Expr
  value: Expr
}

export interface VarDeclaration extends Stmt {
  kind: "VarDeclaration"
  modifier: VarModifier
  identifier: string
  type?: Type
  value?: Expr
}

export interface MultiVarDeclaration extends Stmt {
  kind: "MultiVarDeclaration"
  variables: VarDeclaration[]
}

export interface FunctionParam extends Stmt {
  kind: "FunctionParam"
  name: string
  type?: Type
  default?: Expr // to assign a default value
}

export interface FunctionDeclaration extends Stmt {
  kind: "FunctionDeclaration"
  name: string
  parameters: FunctionParam[]
  body: Stmt[]
}

export interface ReturnStatement extends Stmt {
  kind: "ReturnStatement"
  value: Expr
}

export interface IfElseStatement extends Stmt {
  kind: "IfElseStatement"
  check: Expr
  body: Stmt[]
  childChecks?: IfElseStatement[] // "else if" checks
  else?: Stmt[]
}

export interface WhileStatement extends Stmt {
  kind: "WhileStatement"
  check: Expr
  body: Stmt[]
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface Expr extends Stmt {}

export interface BinaryExpr extends Expr {
  kind: "BinaryExpr"
  left: Expr
  right: Expr
  operator: string
}

export interface MemberExpr extends Expr {
  kind: "MemberExpr"
  object: Expr
  property: Expr
  computed: boolean
}

export interface CallExpr extends Expr {
  kind: "CallExpr"
  args: Expr[]
  caller: Expr
}

export interface Identifier extends Expr {
  kind: "Identifier"
  symbol: string
}

export interface NumericLiteral extends Expr {
  kind: "NumericLiteral"
  value: number
}

export interface StringLiteral extends Expr {
  kind: "StringLiteral"
  value: string
  identifiers: string[] // handle embeded variables
  expressions: { [key: string]: Expr }
}

export interface Property extends Expr {
  kind: "Property"
  key: string
  value?: Expr // its optional to support shorthand i.e. {key}
}

export interface ObjectLiteral extends Expr {
  kind: "ObjectLiteral"
  properties: Property[]
}

export interface ArrayLiteral extends Expr {
  kind: "ArrayLiteral"
  elements: []
}

export interface DocComment extends Expr {
  kind: "DocComment"
}
