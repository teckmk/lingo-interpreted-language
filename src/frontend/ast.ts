export type NodeType =
  // Statements
  | "Program"
  | "VarDeclaration"
  | "MultiVarDeclaration"
  | "FunctionDeclaration"
  | "ReturnStatement"
  | "IfElseStatement"
  | "WhileStatement"
  | "ForStatement"
  | "ForInStatement"
  | "ForRangeStatement"
  | "BreakStatement"
  | "ContinueStatement"
  | "TypeDeclaration"
  | "ContractFulfillment"
  // Expressions
  | "AssignmentExpr"
  | "MemberExpr"
  | "CallExpr"
  | "BinaryExpr"
  | "FunctionParam"
  | "StructMember"
  // Literals
  | "Property"
  | "ObjectLiteral"
  | "StringLiteral"
  | "NumericLiteral"
  | "Identifier"
  | "ArrayLiteral"
  | "DocComment"
  | "SelfKeyword"
  // Types
  | "TypeDeclaration"
  | "AliasDeclaration"
  | "PrimitiveType"
  | "StructType"
  | "AliasType"
  | "GenericType"
  | "UnionType"
  | "ArrayType"
  | "TypeParameter"
  | "FunctionType"
  | "GetterType"
  | "ContractType"

export type Type = "string" | "number" | "bool" | "array" | "object" | "dynamic"
export type VarModifier = "constant" | "final" | "variable"

export interface LeafNode<T> {
  value: T
  position: {
    start: { line: number; column: number }
    end: { line: number; column: number }
  }
}

export interface Stmt {
  kind: NodeType
}

export interface Program extends Stmt {
  kind: "Program"
  body: Stmt[]
}

export interface VarDeclaration extends Stmt {
  kind: "VarDeclaration"
  modifier: VarModifier
  identifier: LeafNode<string>
  type?: TypeNode
  value?: Expr
}

export interface MultiVarDeclaration extends Stmt {
  kind: "MultiVarDeclaration"
  variables: VarDeclaration[]
}

export interface FunctionParam extends Stmt {
  kind: "FunctionParam"
  name: LeafNode<string>
  type?: TypeNode
  default?: Expr // to assign a default value
}

export interface FunctionDeclaration extends Stmt {
  kind: "FunctionDeclaration"
  name: LeafNode<string>
  parameters: FunctionParam[]
  returnType: TypeNode | TypeNode[] | undefined
  body: Stmt[]
  typeParameters?: TypeParameter[] // to support generic type parameters
}

export interface CallExpr extends Expr {
  kind: "CallExpr"
  args: Expr[]
  caller: Expr
  typeArgs?: TypeNode[] // Add support for type arguments in function calls
}

export interface ReturnStatement extends Stmt {
  kind: "ReturnStatement"
  value: Expr | Expr[] // to support multiple return values
}

export interface IfElseStatement extends Stmt {
  kind: "IfElseStatement"
  condition: Expr
  body: Stmt[]
  branches?: IfElseStatement[] // "else if" checks
  else?: Stmt[]
}

export interface WhileStatement extends Loop {
  kind: "WhileStatement"
  check: Expr
  body: Stmt[]
}

export interface Loop extends Stmt {
  loopId: string
  label?: LeafNode<string>
  body: Stmt[]
}

export interface ForStatement extends Loop {
  kind: "ForStatement"
  initializer?: Stmt
  condition?: Expr
  update?: Stmt
}

export interface ForInStatement extends Loop {
  kind: "ForInStatement"
  valueIdentifier: LeafNode<string>
  indexIdentifier: LeafNode<string>
  iterable: Expr
}

export interface ForRangeStatement extends Loop {
  kind: "ForRangeStatement"
  valueIdentifier: LeafNode<string>
  indexIdentifier: LeafNode<string>
  start: Expr
  end?: Expr
  step?: Expr
  inclusive: boolean // false for "to" and true for "through"
}

export interface BreakStatement extends Stmt {
  kind: "BreakStatement"
  loopId: string
  label?: LeafNode<string>
  value?: Expr // to support breaking with a value
}

export interface ContinueStatement extends Stmt {
  kind: "ContinueStatement"
  loopId: string
  label?: LeafNode<string>
}

export interface ContractFulfillment extends Stmt {
  kind: "ContractFulfillment"
  contract?: LeafNode<string> // undefined for default contract
  typeArgs?: TypeNode[] // Add support for type arguments in contract fulfillment
  for: LeafNode<string>
  members: (GetterType | FunctionType)[]
}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface Expr extends Stmt {}

export interface BinaryExpr extends Expr {
  kind: "BinaryExpr"
  left: Expr
  right: Expr
  operator: LeafNode<string>
}

export interface AssignmentExpr extends Expr {
  kind: "AssignmentExpr"
  assigne: Expr
  value: Expr
}

export interface MemberExpr extends Expr {
  kind: "MemberExpr"
  object: Expr
  property: Expr
  computed: boolean
}

// instance of struct
export interface ObjectLiteral extends Expr {
  kind: "ObjectLiteral"
  instanceOf: LeafNode<string>
  properties: Property[]
  typeArgs?: TypeNode[] // Add support for type arguments in object instantiation
}

export interface Identifier extends Expr {
  kind: "Identifier"
  symbol: LeafNode<string>
}

export interface SelfKeyword extends Expr {
  kind: "SelfKeyword"
  symbol: LeafNode<string>
}

export interface NumericLiteral extends Expr {
  kind: "NumericLiteral"
  value: LeafNode<number>
}

export interface StringLiteral extends Expr {
  kind: "StringLiteral"
  value: LeafNode<string>
  identifiers: string[] // handle embeded variables
  expressions: Record<string, Expr> // handle embeded expressions
}

export interface Property extends Expr {
  kind: "Property"
  key: LeafNode<string>
  value?: Expr // its optional to support shorthand i.e. {key}
}

export interface ArrayLiteral extends Expr {
  kind: "ArrayLiteral"
  elements: []
}

export interface DocComment extends Expr {
  kind: "DocComment"
}

/**
 * ==========
 * Types AST
 * ==========
 */

export interface TypeNode extends Expr {
  isNominal: boolean
  name?: LeafNode<string>
}

// 'alias' keyword is for creating aliases to existing types
// ie
// alias User = Response
// User == Response // true
export interface AliasDeclaration extends Stmt {
  kind: "AliasDeclaration"
  name: TypeNode
  type: TypeNode

  // we need this to differentiate b/w generic type:
  // alias User<T> = Response<T>
  // and
  // alias UserR = Response<User> // note: UserR is not generic, on left
  parameters?: TypeNode[]

  isNominal: false // Added this flag just for clarity
}

// 'type' keyword is for distinct types
// ie
// type User Response
// User == Response // false
// User == User // true
export interface TypeDeclaration extends Stmt {
  kind: "TypeDeclaration"
  name: TypeNode
  type: TypeNode

  // we need this to differentiate b/w generic type:
  // type User<T> Response<T>
  // and
  // type UserR Response<User> // note: UserR is not generic, on left
  parameters?: TypeNode[]

  isNominal: true
}

export interface ContractType extends TypeNode {
  kind: "ContractType"
  name?: LeafNode<string>

  members: (GetterType | FunctionType)[]
}

export interface PrimitiveType extends TypeNode {
  kind: "PrimitiveType"
  name: LeafNode<"string" | "number" | "bool" | "dynamic" | "void">
}

export interface StructType extends TypeNode {
  kind: "StructType"
  name?: LeafNode<string>
  members: StructMember[]
}

export interface StructMember extends TypeNode {
  kind: "StructMember"
  name: LeafNode<string>
  type: TypeNode // Supports structs, generics, etc.
  optional: boolean
}

export interface AliasType extends TypeNode {
  kind: "AliasType"
  name?: LeafNode<string>
  actualType: TypeNode
}

export interface GenericType extends TypeNode {
  kind: "GenericType"
  name?: LeafNode<string>
  parameters: TypeParameter[] // Changed from TypeNode[] to TypeParameter[]
}

// New interface to represent type parameters with constraints
export interface TypeParameter extends Expr {
  kind: "TypeParameter"
  name: LeafNode<string>
  constraint?: TypeNode // Optional constraint (e.g., "number" in T: number)
}

export interface UnionType extends TypeNode {
  kind: "UnionType"
  name?: LeafNode<string>
  types: TypeNode[]
}

export interface ArrayType extends TypeNode {
  kind: "ArrayType"
  name?: LeafNode<string>
  elementType: TypeNode
}

export interface FunctionType extends TypeNode {
  kind: "FunctionType"
  name?: LeafNode<string>
  parameters: FunctionParam[]
  returnType: TypeNode | TypeNode[]
  typeParameters?: TypeParameter[]
  body?: Stmt[] // undefined in contracts, and defined in contract fulfillment
}

// ie get name -> string
export interface GetterType extends TypeNode {
  kind: "GetterType"
  name: LeafNode<string>
  parameters?: FunctionParam[] // for self keyword in contract
  returnType: TypeNode | TypeNode[]
  body?: Stmt[]
}
