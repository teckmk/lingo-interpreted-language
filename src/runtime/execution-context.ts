export type StackFrame = {
  functionName: string
  filename: string
  position: {
    start: { line: number; column: number }
    end: { line: number; column: number }
  }
}

export class ExecutionContext {
  private callStack: StackFrame[]

  constructor() {
    this.callStack = []
  }

  public pushCallStack(frame: StackFrame) {
    this.callStack.push(frame)
  }

  public popCallStack() {
    this.callStack.pop()
  }

  public getCallStack(): StackFrame[] {
    return this.callStack.reverse()
  }

  static formatStackTrace(stack: StackFrame[], maxFrames = 10): string {
    const trimmedStack = stack.slice(0, maxFrames) // Limit frames to avoid excessive output
    let stackTrace = "Stack Trace:\n"
    for (const frame of trimmedStack) {
      // const shortFile = frame.filename.replace(/^.*[\\/]/, "") // Extract filename from path
      const shortFile = frame.filename
      stackTrace += `  at ${frame.functionName || "<anonymous>"} (${shortFile}:${frame.position.start.line}:${frame.position.start.column})\n`
    }

    if (stack.length > maxFrames) {
      stackTrace += `  ... (${stack.length - maxFrames} more frames)\n`
    }

    return stackTrace
  }
}
