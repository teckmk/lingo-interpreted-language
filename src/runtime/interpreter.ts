import Environment from "./environment"
import { NumberVal, RuntimeVal, StringVal } from "./values"
import {
  ArrayLiteral,
  AssignmentExpr,
  BinaryExpr,
  CallExpr,
  DocComment,
  FunctionDeclaration,
  Identifier,
  IfElseStatement,
  MemberExpr,
  MultiVarDeclaration,
  NumericLiteral,
  ObjectLiteral,
  Program,
  ReturnStatement,
  Stmt,
  StringLiteral,
  VarDeclaration,
  WhileStatement,
} from "../frontend/2-ast"

import {
  eval_fn_declaration,
  eval_if_else_statement,
  eval_multi_var_declaration,
  eval_program,
  eval_return_statement,
  eval_var_declaration,
  eval_while_statement,
} from "./eval/statements"
import {
  eval_array_expr,
  eval_assignment,
  eval_binary_expr,
  eval_call_expr,
  eval_comment_expr,
  eval_identifier,
  eval_member_expr,
  eval_object_expr,
  eval_string_literal,
} from "./eval/expression"

export function evaluate(astNode: Stmt, env: Environment): RuntimeVal {
  switch (astNode.kind) {
    case "NumericLiteral":
      return {
        value: (astNode as NumericLiteral).value,
        type: "number",
      } as NumberVal
    case "StringLiteral":
      return eval_string_literal(astNode as StringLiteral, env) as StringVal
    case "BinaryExpr":
      return eval_binary_expr(astNode as BinaryExpr, env)
    case "AssignmentExpr":
      return eval_assignment(astNode as AssignmentExpr, env)
    case "Identifier":
      return eval_identifier(astNode as Identifier, env)
    case "ObjectLiteral":
      return eval_object_expr(astNode as ObjectLiteral, env)
    case "ArrayLiteral":
      return eval_array_expr(astNode as ArrayLiteral, env)
    case "CallExpr":
      return eval_call_expr(astNode as CallExpr, env)
    case "Program":
      return eval_program(astNode as Program, env)
    case "VarDeclaration":
      return eval_var_declaration(astNode as VarDeclaration, env)
    case "MultiVarDeclaration":
      return eval_multi_var_declaration(astNode as MultiVarDeclaration, env)
    case "FunctionDeclaration":
      return eval_fn_declaration(astNode as FunctionDeclaration, env)
    case "ReturnStatement":
      return eval_return_statement(astNode as ReturnStatement, env)
    case "IfElseStatement":
      return eval_if_else_statement(astNode as IfElseStatement, env)
    case "WhileStatement":
      return eval_while_statement(astNode as WhileStatement, env)
    case "MemberExpr":
      return eval_member_expr(astNode as MemberExpr, env)
    case "DocComment":
      return eval_comment_expr(astNode as DocComment, env)
    default:
      console.error("This AST can't be interpreted! For now atleast!", astNode)
      process.exit(0)
  }
}
