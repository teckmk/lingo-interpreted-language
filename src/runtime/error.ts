import { ExecutionContext } from "./execution-context"

export class RuntimeError extends Error {
  context?: ExecutionContext

  constructor(context: ExecutionContext, message: string) {
    super(message)
    this.name = "RuntimeError"
    this.context = context
  }
}
