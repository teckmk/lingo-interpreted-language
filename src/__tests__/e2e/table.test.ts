import { parse } from "../../frontend/parser";
import { interpret } from "../../runtime/interpreter";

describe("Table - E2E", () => {
  it("should interpret table of 5", () => {
    const code = `
fn printTable(n: number) -> string {
    let mut table: string = ""
    for let mut i,v in range 1 to 10 {
        let result = n * v
        table = table + "$n x $v = $result" + "\n"
    }

    return table
}

printTable(5)

`;

    const ast = parse("test", code);
    expect(ast).toMatchSnapshot();

    const val = interpret("test", code);
    expect(val).toEqual({
      returned: true,
      type: "return",
      value: {
        returned: false,
        type: "string",
        value: `5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
`,
      },
    });
  });

  it("should interpret table of 5, with inclusive range", () => {
    const code = `
fn printTable(n: number) -> string {
    let mut table: string = ""
    for let mut i,v in range 1 through 10 {
        let result = n * v
        table = table + "$n x $v = $result" + "\n"
    }

    return table
}

printTable(5)

`;

    const ast = parse("test", code);
    expect(ast).toMatchSnapshot();

    const val = interpret("test", code);
    expect(val).toEqual({
      returned: true,
      type: "return",
      value: {
        returned: false,
        type: "string",
        value: `5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50
`,
      },
    });
  });

  it("should interpret for range loop over array", () => {
    const code = `
    let mut arr = [1, 2, 3, 4, 5]
    let mut sum = 0

    for let mut i,v in arr {
        sum = sum + v
    }

    sum
    `;

    const val = interpret("test", code);

    expect(val).toEqual({
      returned: false,
      type: "number",
      value: 15,
    });
  });

  it("should interpret for range loop over string", () => {
    const code = `
    let mut str = "hello"
    let mut result = ""

    for let mut i,v in str {
        result = result + " " + v
    }

    result
    `;

    const val = interpret("test", code);

    expect(val).toEqual({
      returned: false,
      type: "string",
      value: " h e l l o",
    });
  });

  it("should interpret for range loop over string with direct assignment", () => {
    // This is how variable mutability works when using for loop
    // 1. Interpreter creates a temporary variable for the loop (inside the loop scope)
    // 2. Then evaluates the loop
    // 3. Interpreter creates a new variable in parent scope with the same name
    const code = `
    let result = for let mut i,v in "hello" {
        result = result + " " + v
    }

    result
    `;

    const val = interpret("test", code);

    expect(val).toEqual({
      returned: false,
      type: "string",
      value: " h e l l o",
    });
  });

  it("should skip iteration if skip keyword is used", () => {
    const code = `
    let result = for let mut i,v in "hello" {
        if (v == "h") {
            skip
        }
        result = result + " " + v
    }

    result
    `;

    const val = interpret("test", code);

    expect(val).toEqual({
      returned: false,
      type: "string",
      value: " e l l o",
    });
  });

  it("should skip iteration if break keyword is used", () => {
    const code = `
    let result = ""
    
    for let mut i,v in "hello" {
        if (v == "e") {
            break
        }
        result = result + " " + v
    }

    result
    `;

    const val = interpret("test", code);

    expect(val).toEqual({
      returned: false,
      type: "string",
      value: " h",
    });
  });
});
