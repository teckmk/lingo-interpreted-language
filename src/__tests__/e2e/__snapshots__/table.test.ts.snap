// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table - E2E should interpret table of 5 1`] = `
{
  "body": [
    {
      "body": [
        {
          "identifier": {
            "position": {
              "end": {
                "column": 18,
                "line": 3,
              },
              "start": {
                "column": 13,
                "line": 3,
              },
            },
            "value": "table",
          },
          "kind": "VarDeclaration",
          "modifier": "variable",
          "type": {
            "isNominal": false,
            "kind": "PrimitiveType",
            "name": {
              "position": {
                "end": {
                  "column": 26,
                  "line": 3,
                },
                "start": {
                  "column": 20,
                  "line": 3,
                },
              },
              "value": "string",
            },
          },
          "value": {
            "expressions": {},
            "identifiers": [],
            "kind": "StringLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 31,
                  "line": 3,
                },
                "start": {
                  "column": 29,
                  "line": 3,
                },
              },
              "value": "",
            },
          },
        },
        {
          "body": [
            {
              "identifier": {
                "position": {
                  "end": {
                    "column": 19,
                    "line": 5,
                  },
                  "start": {
                    "column": 13,
                    "line": 5,
                  },
                },
                "value": "result",
              },
              "kind": "VarDeclaration",
              "modifier": "final",
              "type": undefined,
              "value": {
                "kind": "BinaryExpr",
                "left": {
                  "kind": "Identifier",
                  "symbol": {
                    "position": {
                      "end": {
                        "column": 23,
                        "line": 5,
                      },
                      "start": {
                        "column": 22,
                        "line": 5,
                      },
                    },
                    "value": "n",
                  },
                },
                "operator": {
                  "position": {
                    "end": {
                      "column": 25,
                      "line": 5,
                    },
                    "start": {
                      "column": 24,
                      "line": 5,
                    },
                  },
                  "value": "*",
                },
                "right": {
                  "kind": "Identifier",
                  "symbol": {
                    "position": {
                      "end": {
                        "column": 27,
                        "line": 5,
                      },
                      "start": {
                        "column": 26,
                        "line": 5,
                      },
                    },
                    "value": "v",
                  },
                },
              },
            },
            {
              "assigne": {
                "kind": "Identifier",
                "symbol": {
                  "position": {
                    "end": {
                      "column": 14,
                      "line": 6,
                    },
                    "start": {
                      "column": 9,
                      "line": 6,
                    },
                  },
                  "value": "table",
                },
              },
              "kind": "AssignmentExpr",
              "value": {
                "kind": "BinaryExpr",
                "left": {
                  "kind": "BinaryExpr",
                  "left": {
                    "kind": "Identifier",
                    "symbol": {
                      "position": {
                        "end": {
                          "column": 22,
                          "line": 6,
                        },
                        "start": {
                          "column": 17,
                          "line": 6,
                        },
                      },
                      "value": "table",
                    },
                  },
                  "operator": {
                    "position": {
                      "end": {
                        "column": 24,
                        "line": 6,
                      },
                      "start": {
                        "column": 23,
                        "line": 6,
                      },
                    },
                    "value": "+",
                  },
                  "right": {
                    "expressions": {},
                    "identifiers": [
                      "$n",
                      "$v",
                      "$result",
                    ],
                    "kind": "StringLiteral",
                    "value": {
                      "position": {
                        "end": {
                          "column": 44,
                          "line": 6,
                        },
                        "start": {
                          "column": 25,
                          "line": 6,
                        },
                      },
                      "value": "$n x $v = $result",
                    },
                  },
                },
                "operator": {
                  "position": {
                    "end": {
                      "column": 46,
                      "line": 6,
                    },
                    "start": {
                      "column": 45,
                      "line": 6,
                    },
                  },
                  "value": "+",
                },
                "right": {
                  "expressions": {},
                  "identifiers": [],
                  "kind": "StringLiteral",
                  "value": {
                    "position": {
                      "end": {
                        "column": 50,
                        "line": 6,
                      },
                      "start": {
                        "column": 47,
                        "line": 6,
                      },
                    },
                    "value": "
",
                  },
                },
              },
            },
          ],
          "end": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 37,
                  "line": 4,
                },
                "start": {
                  "column": 35,
                  "line": 4,
                },
              },
              "value": 10,
            },
          },
          "inclusive": false,
          "indexIdentifier": {
            "position": {
              "end": {
                "column": 18,
                "line": 4,
              },
              "start": {
                "column": 17,
                "line": 4,
              },
            },
            "type": "IDENTIFIER",
            "value": "i",
          },
          "kind": "ForRangeStatement",
          "label": undefined,
          "loopId": "loop_1",
          "start": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 31,
                  "line": 4,
                },
                "start": {
                  "column": 30,
                  "line": 4,
                },
              },
              "value": 1,
            },
          },
          "step": undefined,
          "valueIdentifier": {
            "position": {
              "end": {
                "column": 20,
                "line": 4,
              },
              "start": {
                "column": 19,
                "line": 4,
              },
            },
            "type": "IDENTIFIER",
            "value": "v",
          },
        },
        {
          "kind": "ReturnStatement",
          "value": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 17,
                  "line": 9,
                },
                "start": {
                  "column": 12,
                  "line": 9,
                },
              },
              "value": "table",
            },
          },
        },
      ],
      "kind": "FunctionDeclaration",
      "name": {
        "position": {
          "end": {
            "column": 14,
            "line": 2,
          },
          "start": {
            "column": 4,
            "line": 2,
          },
        },
        "value": "printTable",
      },
      "parameters": [
        {
          "kind": "FunctionParam",
          "name": {
            "position": {
              "end": {
                "column": 16,
                "line": 2,
              },
              "start": {
                "column": 15,
                "line": 2,
              },
            },
            "value": "n",
          },
          "type": {
            "isNominal": false,
            "kind": "PrimitiveType",
            "name": {
              "position": {
                "end": {
                  "column": 24,
                  "line": 2,
                },
                "start": {
                  "column": 18,
                  "line": 2,
                },
              },
              "value": "number",
            },
          },
        },
      ],
      "returnType": {
        "isNominal": false,
        "kind": "PrimitiveType",
        "name": {
          "position": {
            "end": {
              "column": 35,
              "line": 2,
            },
            "start": {
              "column": 29,
              "line": 2,
            },
          },
          "value": "string",
        },
      },
      "typeParameters": undefined,
    },
    {
      "args": [
        {
          "kind": "NumericLiteral",
          "value": {
            "position": {
              "end": {
                "column": 13,
                "line": 12,
              },
              "start": {
                "column": 12,
                "line": 12,
              },
            },
            "value": 5,
          },
        },
      ],
      "caller": {
        "kind": "Identifier",
        "symbol": {
          "position": {
            "end": {
              "column": 11,
              "line": 12,
            },
            "start": {
              "column": 1,
              "line": 12,
            },
          },
          "value": "printTable",
        },
      },
      "kind": "CallExpr",
    },
  ],
  "kind": "Program",
}
`;

exports[`Table - E2E should interpret table of 5, with inclusive range 1`] = `
{
  "body": [
    {
      "body": [
        {
          "identifier": {
            "position": {
              "end": {
                "column": 18,
                "line": 3,
              },
              "start": {
                "column": 13,
                "line": 3,
              },
            },
            "value": "table",
          },
          "kind": "VarDeclaration",
          "modifier": "variable",
          "type": {
            "isNominal": false,
            "kind": "PrimitiveType",
            "name": {
              "position": {
                "end": {
                  "column": 26,
                  "line": 3,
                },
                "start": {
                  "column": 20,
                  "line": 3,
                },
              },
              "value": "string",
            },
          },
          "value": {
            "expressions": {},
            "identifiers": [],
            "kind": "StringLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 31,
                  "line": 3,
                },
                "start": {
                  "column": 29,
                  "line": 3,
                },
              },
              "value": "",
            },
          },
        },
        {
          "body": [
            {
              "identifier": {
                "position": {
                  "end": {
                    "column": 19,
                    "line": 5,
                  },
                  "start": {
                    "column": 13,
                    "line": 5,
                  },
                },
                "value": "result",
              },
              "kind": "VarDeclaration",
              "modifier": "final",
              "type": undefined,
              "value": {
                "kind": "BinaryExpr",
                "left": {
                  "kind": "Identifier",
                  "symbol": {
                    "position": {
                      "end": {
                        "column": 23,
                        "line": 5,
                      },
                      "start": {
                        "column": 22,
                        "line": 5,
                      },
                    },
                    "value": "n",
                  },
                },
                "operator": {
                  "position": {
                    "end": {
                      "column": 25,
                      "line": 5,
                    },
                    "start": {
                      "column": 24,
                      "line": 5,
                    },
                  },
                  "value": "*",
                },
                "right": {
                  "kind": "Identifier",
                  "symbol": {
                    "position": {
                      "end": {
                        "column": 27,
                        "line": 5,
                      },
                      "start": {
                        "column": 26,
                        "line": 5,
                      },
                    },
                    "value": "v",
                  },
                },
              },
            },
            {
              "assigne": {
                "kind": "Identifier",
                "symbol": {
                  "position": {
                    "end": {
                      "column": 14,
                      "line": 6,
                    },
                    "start": {
                      "column": 9,
                      "line": 6,
                    },
                  },
                  "value": "table",
                },
              },
              "kind": "AssignmentExpr",
              "value": {
                "kind": "BinaryExpr",
                "left": {
                  "kind": "BinaryExpr",
                  "left": {
                    "kind": "Identifier",
                    "symbol": {
                      "position": {
                        "end": {
                          "column": 22,
                          "line": 6,
                        },
                        "start": {
                          "column": 17,
                          "line": 6,
                        },
                      },
                      "value": "table",
                    },
                  },
                  "operator": {
                    "position": {
                      "end": {
                        "column": 24,
                        "line": 6,
                      },
                      "start": {
                        "column": 23,
                        "line": 6,
                      },
                    },
                    "value": "+",
                  },
                  "right": {
                    "expressions": {},
                    "identifiers": [
                      "$n",
                      "$v",
                      "$result",
                    ],
                    "kind": "StringLiteral",
                    "value": {
                      "position": {
                        "end": {
                          "column": 44,
                          "line": 6,
                        },
                        "start": {
                          "column": 25,
                          "line": 6,
                        },
                      },
                      "value": "$n x $v = $result",
                    },
                  },
                },
                "operator": {
                  "position": {
                    "end": {
                      "column": 46,
                      "line": 6,
                    },
                    "start": {
                      "column": 45,
                      "line": 6,
                    },
                  },
                  "value": "+",
                },
                "right": {
                  "expressions": {},
                  "identifiers": [],
                  "kind": "StringLiteral",
                  "value": {
                    "position": {
                      "end": {
                        "column": 50,
                        "line": 6,
                      },
                      "start": {
                        "column": 47,
                        "line": 6,
                      },
                    },
                    "value": "
",
                  },
                },
              },
            },
          ],
          "end": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 42,
                  "line": 4,
                },
                "start": {
                  "column": 40,
                  "line": 4,
                },
              },
              "value": 10,
            },
          },
          "inclusive": true,
          "indexIdentifier": {
            "position": {
              "end": {
                "column": 18,
                "line": 4,
              },
              "start": {
                "column": 17,
                "line": 4,
              },
            },
            "type": "IDENTIFIER",
            "value": "i",
          },
          "kind": "ForRangeStatement",
          "label": undefined,
          "loopId": "loop_1",
          "start": {
            "kind": "NumericLiteral",
            "value": {
              "position": {
                "end": {
                  "column": 31,
                  "line": 4,
                },
                "start": {
                  "column": 30,
                  "line": 4,
                },
              },
              "value": 1,
            },
          },
          "step": undefined,
          "valueIdentifier": {
            "position": {
              "end": {
                "column": 20,
                "line": 4,
              },
              "start": {
                "column": 19,
                "line": 4,
              },
            },
            "type": "IDENTIFIER",
            "value": "v",
          },
        },
        {
          "kind": "ReturnStatement",
          "value": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 17,
                  "line": 9,
                },
                "start": {
                  "column": 12,
                  "line": 9,
                },
              },
              "value": "table",
            },
          },
        },
      ],
      "kind": "FunctionDeclaration",
      "name": {
        "position": {
          "end": {
            "column": 14,
            "line": 2,
          },
          "start": {
            "column": 4,
            "line": 2,
          },
        },
        "value": "printTable",
      },
      "parameters": [
        {
          "kind": "FunctionParam",
          "name": {
            "position": {
              "end": {
                "column": 16,
                "line": 2,
              },
              "start": {
                "column": 15,
                "line": 2,
              },
            },
            "value": "n",
          },
          "type": {
            "isNominal": false,
            "kind": "PrimitiveType",
            "name": {
              "position": {
                "end": {
                  "column": 24,
                  "line": 2,
                },
                "start": {
                  "column": 18,
                  "line": 2,
                },
              },
              "value": "number",
            },
          },
        },
      ],
      "returnType": {
        "isNominal": false,
        "kind": "PrimitiveType",
        "name": {
          "position": {
            "end": {
              "column": 35,
              "line": 2,
            },
            "start": {
              "column": 29,
              "line": 2,
            },
          },
          "value": "string",
        },
      },
      "typeParameters": undefined,
    },
    {
      "args": [
        {
          "kind": "NumericLiteral",
          "value": {
            "position": {
              "end": {
                "column": 13,
                "line": 12,
              },
              "start": {
                "column": 12,
                "line": 12,
              },
            },
            "value": 5,
          },
        },
      ],
      "caller": {
        "kind": "Identifier",
        "symbol": {
          "position": {
            "end": {
              "column": 11,
              "line": 12,
            },
            "start": {
              "column": 1,
              "line": 12,
            },
          },
          "value": "printTable",
        },
      },
      "kind": "CallExpr",
    },
  ],
  "kind": "Program",
}
`;
